### DISCUSSION TITLE
How to inject to open devtools in Electron

### DISCUSSION THREAD

**Discussion by CHENKAI00** [ShellWen/v8_killer#21] [Q&A]
I successfully injected your config. But But I can't see any effect. Where are the parameters of “console.log ” logged? In addition, This project seems to only support injecting native v8 functions, which I am not sure. I have a question. If I want to open devtools, What code should I inject instead? I tried this script:
`var allWebContents = require('electron').webContents.getAllWebContents()
allWebContents.forEach((contents)=>{
    if (contents.openDevTools) {
        contents.openDevTools();
    }
}` But failed.
I don't know if my way is wrong. I sincerely look forward to your reply. Thank you very much.

**Comment by ShellWen** [comment]
It will log to stdout default. Could you provide log?

**Comment by CHENKAI00** [comment]
**E:\app\v8_killer-master\target\debug>v8_killer_launcher.exe E:\bilibili-debugger\bilibili\bi.exe
[*] Executable: E:\bilibili-debugger\bilibili\bi.exe
[*] Args: []
[*] Core lib path: E:\app\v8_killer-master\target\debug\v8_killer_core.dll
[*] Creating process.
[*] PID: 16900
[*] Alloc core lib path memory.
[*] Remote lib path memory Address: 0x1f7bca20000.
[*] Writing core lib path to process.
[*] Getting LoadLibraryW address.
[*] Creating remote thread.
[*] Core lib inject success. Waiting for thread end.
[*] V8_KILLER_CONFIG_FILE_PATH: E:\app\v8_killer-master\examples\configs\simple.toml
[*] Read Config success
[*] Config: Config { rules: {"hook_console_log": ConfigRule { matcher: ResourceNameKeywordMatcher(ResourceNameKeywordMatcher { keyword: "hello_world.js" }), processors: [InsertBeforeProcessor(InsertBeforeProcessor { content: "const originalConsoleLog = console.log\n    console.log = function (x) {\n\t\tdebugger;\n\t\t\n      originalConsoleLog('now you see :)',x)\n      return originalConsoleLog.apply(this, arguments)\n    }" }), InsertAfterProcessor(InsertAfterProcessor { content: "console.log = originalConsoleLog;\n\n    " })] }} }
[*] v8_script_compiler_compile_function_internal found: 0x7ff60524c0a0
[*] Thread ended. Resume original thread.
[*] --- Following is the original process output ---

21:17:34.331 (AppIndex) > Starting...
21:17:34.353 (AppIndex) > Process args ["E:\\bilibili-debugger\\bilibili\\bi.exe"]
21:17:34.550 (AppIndex) > Creating new instance!**
This is the log. In your config file, You replaced “originalConsoleLog” with “console.log” and added 'now you see:' log. But I can not see it. The hook script seems to have no effect.

**Comment by CHENKAI00** [comment]
Sorry, I seem to know where I am wrong

**Comment by CHENKAI00** [comment]
But I still do not konw what script I can inject to open devtools.

**Comment by CHENKAI00** [comment]
In addition, I found some electron applications were started with "jsc" files, and the injection failed again.
I'm very sorry to keep submitting issues to you. I'm sorry if I have disturbed you.

**Reply by ShellWen** [reply]
jsc? can you give me screenshots?

**Reply by z1queue** [reply]
我也是同样的问题，Windows 11 家庭版 22H2  22621.2715  64 位操作系统, 基于 ARM 的处理器
下面是我执行node的console.log("1234")，但是并没有看到hook后的输出now you see :
C:\Users\work\Desktop\v8_killer\target\debug>v8_killer_launcher.exe  "C:\Program Files\nodejs\node.exe"
[*] Executable: C:\Program Files\nodejs\node.exe
[*] Args: []
[*] Core lib path: C:\Users\work\Desktop\v8_killer\target\debug\v8_killer_core.dll
[*] Creating process.
[*] PID: 10332
[*] Alloc core lib path memory.
[*] Remote lib path memory Address: 0x1e3b8420000.
[*] Writing core lib path to process.
[*] Getting LoadLibraryW address.
[*] Creating remote thread.
[*] Core lib inject success. Waiting for thread end.
[*] V8_KILLER_CONFIG_FILE_PATH: C:\Users\work\Desktop\v8_killer\examples\configs\simple.toml
[*] Read Config success
[*] Config: Config { identifiers: Identifiers { V8_SCRIPT_COMPILER_COMPILE_FUNCTION_INTERNAL: [SymbolIdentifier(SymbolIdentifier { symbols: ["_ZN2v814ScriptCompiler23CompileFunctionInternalENS_5LocalINS_7ContextEEEPNS0_6SourceEmPNS1_INS_6StringEEEmPNS1_INS_6ObjectEEENS0_14CompileOptionsENS0_13NoCacheReasonEPNS1_INS_14ScriptOrModuleEEE", "?CompileFunctionInternal@ScriptCompiler@v8@@CA?AV?$MaybeLocal@VFunction@v8@@@2@V?$Local@VContext@v8@@@2@PEAVSource@12@_KQEAV?$Local@VString@v8@@@2@2QEAV?$Local@VObject@v8@@@2@W4CompileOptions@12@W4NoCacheReason@12@PEAV?$Local@VScriptOrModule@v8@@@2@@Z"] })], V8_STRING_UTF8LENGTH: [SymbolIdentifier(SymbolIdentifier { symbols: ["_ZNK2v86String10Utf8LengthEPNS_7IsolateE", "?Utf8Length@String@v8@@QEBAHPEAVIsolate@2@@Z"] })], V8_STRING_WRITE_UTF8: [SymbolIdentifier(SymbolIdentifier { symbols: ["_ZNK2v86String9WriteUtf8EPNS_7IsolateEPciPii", "?WriteUtf8@String@v8@@QEBAHPEAVIsolate@2@PEADHPEAHH@Z"] })], V8_CONTEXT_GET_ISOLATE: [SymbolIdentifier(SymbolIdentifier { symbols: ["_ZN2v87Context10GetIsolateEv", "?GetIsolate@Context@v8@@QEAAPEAVIsolate@2@XZ"] })], V8_STRING_NEW_FROM_UTF8: [SymbolIdentifier(SymbolIdentifier { symbols: ["_ZN2v86String11NewFromUtf8EPNS_7IsolateEPKcNS_13NewStringTypeEi", "?NewFromUtf8@String@v8@@SA?AV?$MaybeLocal@VString@v8@@@2@PEAVIsolate@2@PEBDW4NewStringType@2@H@Z"] })] }, rules: {"hook_console_log": ConfigRule { matcher: ResourceNameKeyword(ResourceNameKeywordMatcher { keyword: "hello_world.js" }), processors: [InsertBefore(InsertBeforeProcessor { content: "const originalConsoleLog = console.log\n    console.log = function () {\n      originalConsoleLog('now you see :)')\n      return originalConsoleLog.apply(this, arguments)\n    }" }), InsertAfter(InsertAfterProcessor { content: "console.log = originalConsoleLog\n    " })] }} }
[*] v8_script_compiler_compile_function_internal found: 0x7ff6f8c4d950
[*] Thread ended. Resume original thread.
[*] --- Following is the original process output ---
Welcome to Node.js v20.9.0.
Type ".help" for more information.
> console.log("1234")
1234
undefined
>

**Comment by oliverstank** [comment]
did you figure this out?
